go/teleportr/drivers/disburser/driver.go

```diff
@@ -0,0 +1,518 @@
+	header, err := d.cfg.L2Client.HeaderByNumber(ctx, receipt.BlockNumber)
+	if err != nil {
+		log.Warn("Unable to fetch block by number",
+			"blockNumber", receipt.BlockNumber,
+			"err", err)
+
+		// Not super critical that this is the actual timestamp, can always
+		// rectify later. Not worth failing over.
+		blockTimestamp = time.Unix(time.Now().Unix(), 0)
+	} else {
+		blockTimestamp = time.Unix(int64(header.Time), 0)
+	}
+
+	for _, depositID := range depositIDs {
+		log.Info("Marking disbursement success",
+			"depositId", depositID,
+			"txHash", receipt.TxHash,
+			"blockNumber", receipt.BlockNumber,
+			"blockTimestamp", blockTimestamp)
+
+		err = d.cfg.Database.UpsertDisbursement(
+			depositID, receipt.TxHash, receipt.BlockNumber.Uint64(),
+			blockTimestamp,
+		)
+		if err != nil {
+			log.Warn("Unable to mark disbursement success",
+				"depositId", depositID,
+				"txHash", receipt.TxHash,
+				"blockNumber", receipt.BlockNumber,
+				"blockTimestamp", blockTimestamp,
+				"err", err)
+			continue
+		}
+
+		log.Info("Disbursement marked success",
+			"depositId", depositID,
+			"txHash", receipt.TxHash,
+			"blockNumber", receipt.BlockNumber,
+			"blockTimestamp", blockTimestamp)
+	}
+
+	return nil
+}
+
+// ingestDeposits is a preprocessing step done each time we attempt to compute a
+// new block range. This method scans for any missing or recent logs from the
+// contract, and upserts any new deposits into the database.
+func (d *Driver) ingestDeposits(
+	ctx context.Context,
+	blockNumber uint64,
+	lastProcessedBlockNumber *uint64,
+) error {
+
+	filterStartBlockNumber := FindFilterStartBlockNumber(
+		FilterStartBlockNumberParams{
+			BlockNumber:              blockNumber,
+			NumConfirmations:         d.cfg.NumConfirmations,
+			DeployBlockNumber:        d.cfg.DeployBlockNumber,
+			LastProcessedBlockNumber: lastProcessedBlockNumber,
+		},
+	)
+
+	// Fetch deposit events in block ranges capped by the FilterQueryMaxBlocks
+	// parameter.
+	maxBlocks := d.cfg.FilterQueryMaxBlocks
+	for start := filterStartBlockNumber; start < blockNumber+1; start += maxBlocks {
+		var end = start + maxBlocks
+		if end > blockNumber+1 {
+			end = blockNumber + 1
+		}
+
+		opts := &bind.FilterOpts{
+			Start:   start,
+			End:     &end,
+			Context: ctx,
+		}
+		events, err := d.depositContract.FilterEtherReceived(opts, nil, nil, nil)
+		if err != nil {
+			return err
+		}
+		defer events.Close()
+
+		var deposits []db.Deposit
+		for events.Next() {
+			event := events.Event
+
+			header, err := d.cfg.L1Client.HeaderByNumber(
+				ctx, big.NewInt(int64(event.Raw.BlockNumber)),
+			)
+			if err != nil {
+				return err
+			}
+
+			deposits = append(deposits, db.Deposit{
+				ID:             event.DepositId.Uint64(),
+				TxnHash:        event.Raw.TxHash,
+				BlockNumber:    event.Raw.BlockNumber,
+				BlockTimestamp: time.Unix(int64(header.Time), 0),
+				Address:        event.Emitter,
+				Amount:         event.Amount,
```

Comment:
There have been hacks in the past where a bridge uses `tx.Value` instead of a value emitted from an `event` which could cause problems if the tx calls a loop, it would count the value multiple times. I believe we are safe from this kind of attack but perhaps its good to add a unit test to the contract tests before this goes into production